File: simplebase/src/main/java/com/zjh/simplebase/manager/ActivityManager.java
package com.zjh.simplebase.manager;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import com.zjh.simplebase.base.BaseActivity;

import java.util.Iterator;
import java.util.Stack;

/**
 * Activity 管理器
 * 需在 Application 中注册「registerActivityLifecycleCallbacks」
 */
public class ActivityManager implements Application.ActivityLifecycleCallbacks {

    private static ActivityManager activityManager;
    private Stack<Activity> activityStack = new Stack<>();

    private ActivityManager() {
    }

    public static ActivityManager getInstance() {
        if (activityManager == null) {
            synchronized (ActivityManager.class) {
                if (activityManager == null) {
                    activityManager = new ActivityManager();
                }
            }
        }
        return activityManager;
    }

    /**
     * Activity 入栈
     */
    private void addActivity(Activity activity) {
        activityStack.add(activity);
    }

    /**
     * Activity 出栈
     */
    private void removeActivity(Activity activity) {
        if (activity != null) {
            activityStack.remove(activity);
        }
    }

    /**
     * 获取当前Activity
     */
    public Activity getCurrentActivity() {
        return activityStack.lastElement();
    }

    /**
     * 获取Activity栈
     */
    public Stack<Activity> getActivityStack() {
        return activityStack;
    }

    /**
     * 结束当前Activity
     */
    public void finishActivity() {
        finishActivity(activityStack.lastElement());
    }

    /**
     * 结束所有Activity
     */
    public void finishAllActivity() {
        for (int i = 0, size = activityStack.size(); i < size; i++) {
            if (null != activityStack.get(i)) {
                activityStack.get(i).finish();
            }
        }
        activityStack.clear();

    }

    /**
     * activity存在
     */
    public boolean hasActivity(Class<? extends BaseActivity> cls) {
        boolean has = false;
        for (Activity activity : activityStack) {
            if (activity.getClass().equals(cls)) {
                has = true;
                break;
            }
        }
        return has;
    }

    /**
     * 结束某 Activity
     */
    public void finishActivity(Activity activity) {
        if (activity != null) {
            removeActivity(activity);
            activity.finish();
        }
    }

    /**
     * 结束某 Activity
     */
    public void finishActivity(Class<? extends BaseActivity> cls) {
        Iterator iterator = activityStack.iterator();
        while (iterator.hasNext()) {
            Activity activity = (Activity) iterator.next();
            if (activity.getClass().equals(cls)) {
                iterator.remove();
                activity.finish();
            }
        }
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        addActivity(activity);
    }

    @Override
    public void onActivityStarted(Activity activity) {

    }

    @Override
    public void onActivityResumed(Activity activity) {

    }

    @Override
    public void onActivityPaused(Activity activity) {

    }

    @Override
    public void onActivityStopped(Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        removeActivity(activity);
    }
}

File: simplebase/src/main/java/com/zjh/simplebase/manager/ActivityManager.kt
package com.zjh.simplebase.manager

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import com.zjh.simplebase.util.SimpleLog
import java.util.concurrent.CopyOnWriteArrayList

/**
 * 界面管理类
 *
 * @author zjh
 * on 2021/4/15
 */
object ActivityManager {

    private const val TAG = "ActivityManager"

    private val list = mutableListOf<Activity>()

    var uiHandler: Handler = Handler(Looper.getMainLooper())
        private set

    fun init(application: Application) {
        application.registerActivityLifecycleCallbacks(object :
                Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
                addActivity(activity)
                printActivityList()
            }

            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
            }

            override fun onActivityStarted(activity: Activity) {
            }

            override fun onActivityResumed(activity: Activity) {
                lastLiteCycleState.resumed()
                switchToForeground(activity)
            }

            override fun onActivityPaused(activity: Activity) {
                lastLiteCycleState.paused()
                checkBackgroundState(activity)
            }

            override fun onActivityStopped(activity: Activity) {
            }

            override fun onActivityDestroyed(activity: Activity) {
                removeActivity(activity)
                printActivityList()
            }

        })
    }

    val curActivity: Activity?
        get() = list.lastOrNull()

    val activitys: List<Activity>
        get() = list.toList()

    /**返回List，存在则List不为空**/
    fun <T : Activity> filter(clazz: Class<T>): List<T> {
        return list.filter {
            it::class.java.name == clazz.name
        }.map { it as T }
    }

    /**
     * 判断是否有某个activity
     */
    fun hasActivity(cls: Class<out Activity?>): Boolean {
        var has = false
        for (activity in list) {
            if (activity.javaClass == cls) {
                has = true
                break
            }
        }
        return has
    }

    /**
     * 关闭某个activity
     */
    fun finishActivity(cls: Class<out Activity?>) {
        for (activity in list) {
            if (activity.javaClass == cls) {
                activity.finish()
                break
            }
        }
    }

    /**关闭所有页面，并在关闭后执行[block]函数**/
    fun finishAll(block: ((activity: Activity?) -> Unit)? = null) {
        val curA = curActivity
        list.forEach {
            it.finish()
        }
        block?.invoke(curA)
    }

    private fun addActivity(activity: Activity) {
        list.add(activity)
    }

    private fun removeActivity(activity: Activity) {
        list.remove(activity)
    }

    private fun printActivityList() {
        SimpleLog.d(TAG) {
            list.fold("") { result, activity ->
                "$result${if (result.isEmpty()) "" else " -> "}${activity::class.java.simpleName}"
            }
        }
    }

    private data class LastLiteCycleState(var isResumed: Boolean, var isPaused: Boolean) {
        fun resumed() {
            isResumed = true
            isPaused = false
        }

        fun paused() {
            isResumed = false
            isPaused = true
        }
    }

    /**是否在前台**/
    var isForeground: Boolean = false
        private set
    private val listeners: MutableList<ForegroundListener> =
            CopyOnWriteArrayList<ForegroundListener>()
    private val lastLiteCycleState: LastLiteCycleState = LastLiteCycleState(
            isResumed = false,
            isPaused = false
    )

    private var runnableBackground: Runnable? = null

    fun addForegroundListener(listener: ForegroundListener) {
        listeners.add(listener)
    }

    fun removeForegroundListener(listener: ForegroundListener) {
        listeners.remove(listener)
    }

    private fun switchToForeground(activity: Activity) {
        val isBackground = !isForeground
        isForeground = true
        runnableBackground?.apply {
            uiHandler.removeCallbacks(this)
        }
        if (isBackground) {
            SimpleLog.d(TAG, "切换到前台：当前页面是${activity::class.java.name}")
            listeners.forEach {
                it.onSwitchToForeground(activity)
            }
        }
    }

    /**检查是否切换到后台**/
    private fun checkBackgroundState(activity: Activity) {
        runnableBackground = Runnable {
            if (isForeground && lastLiteCycleState.isPaused) {
                isForeground = false
                SimpleLog.d(TAG, "切换到后台：从${activity::class.java.name}页面中切换到后台")
                listeners.forEach {
                    it.onSwitchToBackground(activity)
                }
            }
        }
        uiHandler.postDelayed(runnableBackground!!, 500)
    }

    interface ForegroundListener {
        /**当APP从后台切换到前台时被调用**/
        fun onSwitchToForeground(activity: Activity)

        /**当APP从前台切换到后台时被调用**/
        fun onSwitchToBackground(activity: Activity)
    }
}